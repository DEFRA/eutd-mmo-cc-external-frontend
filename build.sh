#/bin/bash
curl -O https://nodejs.org/dist/v12.18.4/node-v12.18.4-linux-x64.tar.xz
tar -xf node-v12.18.4-linux-x64.tar.xz --directory /tmp
export PATH="/tmp/node-v12.18.4-linux-x64/bin:$PATH"
rm node-v12.18.4-linux-x64.tar.xz
npm ci
npm run test-vstack
rm -rf /tmp/node-v12.18.4-linux-x64