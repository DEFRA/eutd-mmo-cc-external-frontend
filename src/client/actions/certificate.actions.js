import { isError, showFullPageError } from './index';

export const GET_SUMMARY_DOCUMENT = 'GET_SUMMARY_DOCUMENT';
export const SHOW_SUMMARY_DOCUMENT_ERRORS = 'SHOW_SUMMARY_DOCUMENT_ERRORS';
export const GET_SUMMARY_UNAUTHORISED = 'GET_SUMMARY_DOCUMENT_UNAUTHORISED';

export function showSummaryCertificateErrors(data) {
  return { type: SHOW_SUMMARY_DOCUMENT_ERRORS, payload: { validationErrors: data } };
}

export const getSummaryCertificate = (journey, documentNumber = undefined) => async (dispatch, getState, { orchestrationApi }) => {
  const config = (documentNumber)
    ? { headers: { documentnumber: documentNumber } }
    : {};

  await orchestrationApi.get(`/certificate/${journey}`, config)
    .then(
      res => {
        dispatch({ type: GET_SUMMARY_DOCUMENT, payload: res.data });
      },
      error => {
        if (isError(error)) {
          dispatch(showFullPageError());
        } else {
          dispatch(unauthorised());
        }
      });

    function unauthorised() { return { type: GET_SUMMARY_UNAUTHORISED }; }
};