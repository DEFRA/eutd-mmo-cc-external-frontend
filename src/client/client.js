// Startup point for the client side application
import 'babel-polyfill';
import React from 'react';
import ReactDOM from 'react-dom';
import {Router, withRouter} from 'react-router-dom';
import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { renderRoutes } from 'react-router-config';
import axios from 'axios';
import Routes from './Routes';
import reducers from './reducers';
import { createBrowserHistory } from 'history';

const composeEnhancers = process.env.NODE_ENV === 'development' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__  ?
  window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ : compose;
  

const paramSerializerFn = params => {
  let output = '';
  Object.keys(params).forEach((k) => {
    output+= `${k}=${params[k]}&`;
  });
  return encodeURI(output);
};

const axiosResponseInterceptorSuccessFn = response => response;
const axiosResponseInterceptorErrorFn = error => {
  if (error.response.status === 599) {
    // Let's check if this is because we're logged out
    const empPromise = new Promise(() => {});
    const searchParam = error.request.responseURL.slice(error.request.responseURL.search(/[?]/g));
    let searchParams = new URLSearchParams(searchParam);
    if(searchParams.get('notLoggedInErr') === 'yes') {
      window.location = '/logout?backToPath=/server-logout';
      return empPromise;
    }
    // It's always a timeout...
    window.location = '/timed-out';
    return new Promise(() => {});
  } else {
    return new Promise((_,rej) => rej(error));
  }
};

const axiosInstanceForOrchService = axios.create({
  baseURL: '/orchestration/api/v1/',
  paramsSerializer: paramSerializerFn,
  timeout: window.INITIAL_STATE.config.axiosTimeout
});

axiosInstanceForOrchService.interceptors.response.use(axiosResponseInterceptorSuccessFn, axiosResponseInterceptorErrorFn);

const axiosInstanceForRefService = axios.create({
  baseURL: '/reference/api/v1/',
  paramsSerializer: paramSerializerFn,
  timeout: window.INITIAL_STATE.config.axiosTimeout
});

axiosInstanceForRefService.interceptors.response.use(axiosResponseInterceptorSuccessFn, axiosResponseInterceptorErrorFn);

const axiosInstanceForDynamixService = axios.create({
  baseURL: '/dynamix/',
  paramsSerializer: paramSerializerFn,
  timeout: window.INITIAL_STATE.config.axiosTimeout
});

axiosInstanceForDynamixService.interceptors.response.use(axiosResponseInterceptorSuccessFn, axiosResponseInterceptorErrorFn);

const store = createStore(
  reducers,
  window.INITIAL_STATE,
  composeEnhancers(applyMiddleware(
    // The difference between orch/ref service api is not significant as they both go through proxy
    // However it is significant in the server side store. To stick with same action creators we're
    // passing down same axios instance to both orchestration and reference service api .
    thunk.withExtraArgument( {
      orchestrationApi: axiosInstanceForOrchService,
      referenceServiceApi: axiosInstanceForRefService,
      dynamixApi: axiosInstanceForDynamixService } )
  ))
);

const ScrollToTop = withRouter( class ScrollToTop extends React.Component {
  componentDidMount() {
  }

  render() {
    return this.props.children;
  }
});

const history = createBrowserHistory();


ReactDOM.hydrate(
  <Provider store={store}>
    <Router history={history}>
      <ScrollToTop>
        <div>{renderRoutes(Routes)}</div>
      </ScrollToTop>
    </Router>
  </Provider>,
  document.querySelector('#root')
);
