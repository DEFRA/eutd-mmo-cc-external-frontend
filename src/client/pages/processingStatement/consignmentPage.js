import React from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';
import BackToProgressLink from '../../components/BackToProgressLink';

import {
  Main,
  BackLink,
  TextArea,
  Header,
  GridRow,
  GridCol
} from 'govuk-react';

import {
  getProcessingStatementFromRedis,
  saveProcessingStatement,
  saveProcessingStatementToRedis,
  clearProcessingStatement
} from '../../actions';

import PageTemplateWrapper from '../../components/PageTemplateWrapper';
import {
  toGovukErrors,
  onHandleErrorClick,
  scrollToErrorIsland
} from '../utils';
import ErrorIsland from '../../components/elements/ErrorIsland';
import PageTitle from '../../components/PageTitle';
import Form from '../../components/elements/Form';
import {withRouter} from 'react-router-dom';
import HelpLink from '../../components/HelpLink';
import SaveAsDraftButton from '../../components/SaveAsDraftButton';
import ContinueButton from '../../components/elements/ContinueButton';
import { withTranslation } from 'react-i18next';
class ConsignmentPage extends React.Component {

  async save(nextUri) {
    const isDescriptionSavedAsDraft = nextUri === this.props.route.saveAsDraftUri;
    const documentNumber = this.props.match.params['documentNumber'];
    const processingStatement = await this.props.saveToRedis({
        data: this.props.processingStatement || {},
        currentUrl: `/create-processing-statement/${documentNumber}/add-consignment-details`,
        saveToRedisIfErrors: true,
        saveAsDraft: isDescriptionSavedAsDraft,
        documentNumber: documentNumber
    });

    if (!isDescriptionSavedAsDraft && !_.isEmpty(processingStatement.errors)) {
      scrollToErrorIsland();
      return;
    }

    this.props.history.push(nextUri.replace(':documentNumber', documentNumber));
  }

  componentWillUnmount() {
    this.props.clear();
  }

  async componentDidMount() {
    const { processingStatement } = this.props;

    await this.props.getFromRedis(this.props.match.params.documentNumber);

    if (processingStatement && processingStatement.unauthorised === true) {
      this.unauthorised();
    }
  }

  unauthorised() {
    this.props.history.push('/forbidden');
  }

  componentDidUpdate() {
    const { processingStatement } = this.props;

    if (processingStatement && processingStatement.unauthorised === true) {
      this.unauthorised();
    }
  }

  onChange(e) {
    const processingStatement = _.cloneDeep(this.props.processingStatement || {});
    processingStatement[e.target.name] = e.target.value;
    this.props.save(processingStatement);
  }

  goBack(e, page) {
    e.preventDefault();
    const processingStatementClone = _.cloneDeep((this.props.processingStatement || {}));
    processingStatementClone.errors = {};
    this.props.save(processingStatementClone);
    this.props.history.push(page);
  }

  onSaveAsDraft = async (e) => {
    e.preventDefault();
    const { saveAsDraftUri } = this.props.route;
    await this.save(saveAsDraftUri);
  }

  render() {
    const { processingStatement = {}, history, route, match, t } = this.props;
    const { previousUri, path, nextUri, saveAsDraftUri, journey, progressUri, title } = route;
    const backUrl = previousUri.replace(':documentNumber', match.params.documentNumber);
    const documentNumber = match.params['documentNumber'];

    const errors = (processingStatement && processingStatement.errorsUrl === history.location.pathname && processingStatement.errors) || {}; // only display errors for this URL
    const saveAsDraftFormActionUrl = `/orchestration/api/v1/processingStatement/saveAndValidate?c=${path}&n=${nextUri}&saveAsDraftUrl=${saveAsDraftUri}`;
    return (
      <Main className="processing-statement">
        <PageTitle title={`${!_.isEmpty(errors) ? 'Error: ' : ''}${t('psAddConsignmentDetailsConsignmentPageHeader')} - ${t(title)}`} />
        <ErrorIsland errors={toGovukErrors(errors)} onHandleErrorClick={onHandleErrorClick}/>
        <GridRow>
          <GridCol>
            <BackLink onClick={e => this.goBack(e, backUrl)} href={backUrl}>
              {t('commonBackLinkBackButtonLabel')}
            </BackLink>
            <Header>{t('psAddConsignmentDetailsConsignmentPageHeader')}</Header>
          </GridCol>
        </GridRow>
        <Form
          action="/orchestration/api/v1/processingStatement/saveAndValidate"
          currentUrl={path}
          nextUrl={nextUri}
          onSubmit={e => e.preventDefault()}>
          <GridRow>
            <GridCol columnTwoThirds>
              <TextArea htmlFor="consignmentDescription" meta={{ error: t(errors.consignmentDescription), touched: true }} input={{autoComplete: 'off', className: 'formControl', name: 'consignmentDescription', id: 'consignmentDescription', value: processingStatement.consignmentDescription, onChange: e => this.onChange(e)}} hint={t('psAddConsignmentDetailsConsignmentDescriptionHint')}></TextArea>
            </GridCol>
          </GridRow>
          <GridRow>
            <SaveAsDraftButton formactionUrl={saveAsDraftFormActionUrl} onClick={this.onSaveAsDraft} />
            <input type="hidden" name="journey" value={journey} />
            <input type="hidden" name="currentUri" value={path} />
            <ContinueButton id="continue" onClick={() => this.save(nextUri)}> {t('commonContinueButtonSaveAndContinueButton')}</ContinueButton>
          </GridRow>
        </Form>
        <BackToProgressLink
          progressUri={progressUri}
          documentNumber={documentNumber}
        />
        <HelpLink journey={journey}/>
      </Main>
    );
  }
}

function mapStateToProps(state) {
  return {
    processingStatement: state.processingStatement
  };
}

function loadData(store) {
  return store.dispatch(getProcessingStatementFromRedis(this.documentNumber));
}

export const component = withRouter(
  connect(mapStateToProps, {
    save: saveProcessingStatement,
    clear: clearProcessingStatement,
    saveToRedis: saveProcessingStatementToRedis,
    getFromRedis: getProcessingStatementFromRedis
   })(withTranslation() (ConsignmentPage)));

export default {
  loadData,
  component: PageTemplateWrapper(component)
};
