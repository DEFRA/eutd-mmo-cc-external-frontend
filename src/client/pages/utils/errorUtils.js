import i18n from '../../../i18n';
import enTranslations from '../../../locales/uk/en.json';

export function toGovukErrors(obj) {
  return Object.keys( obj).map( o => {
    return {targetName: o, text: obj[o], key: o, message: (obj[o] in enTranslations) ? i18n.t(obj[o]) : obj[o]};
  });
}

export function onHandleErrorClick(targetName, speciesRow, vesselIndex) {
  let elem;

  if (speciesRow !== undefined && vesselIndex !== undefined) {
    elem = document.getElementById(targetName + '-' + speciesRow + '-' + vesselIndex);
  } else {
    elem = document.getElementsByName(targetName)[0] || document.getElementById(targetName);
  }

  if (elem) {
    elem.scrollIntoView({ behavior: 'smooth' });
    elem.focus({ preventScroll: true});
  }
}

export function onHandleErrorClickWithoutFocus(targetName) {
  const elem = document.getElementsByName(targetName)[0] || document.getElementById(targetName);
  elem.scrollIntoView({ behavior: 'smooth' });
}

export function scrollToErrorIsland() {
  setTimeout( () => {
    document.getElementById('errorIsland').scrollIntoView({behavior: 'smooth'});
    document.getElementById('errorIsland').focus({ preventScroll: true});
  },100);
}

export function scrollToField(id) {
  setTimeout( () => {
    document.getElementById(id).scrollIntoView({behavior: 'smooth'});
    document.getElementById(id).focus({ preventScroll: true});
  },100);
}

export function scrollToFieldName(name) {
  setTimeout( () => {
    document.getElementsByName(name)[0].scrollIntoView({behavior: 'smooth'});
    document.getElementsByName(name)[0].focus({ preventScroll: true});
  },100);
}

