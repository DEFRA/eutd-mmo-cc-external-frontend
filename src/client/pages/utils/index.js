import React, {Fragment} from 'react';

export * from './errorUtils';


export function formatAddress( ...adresssLines) {
  return adresssLines.filter( a => a ).map( (a, index) => <Fragment key={index}>{a}<br/></Fragment> );
}

export function constructAddress(addressArray) {
  let address = '';
  if (addressArray.length === 4) {
    if (addressArray[0]) {
      address += addressArray[0];
    }
    if (addressArray[0] && addressArray[1]) {
      address += ' ';
    }
    if (addressArray[1]) {
      address += addressArray[1];
    }
    if (addressArray[2]) {
      address += ', ' + addressArray[2];
    }
    if (addressArray[3]) {
      address += '. ' + addressArray[3];
    }
  }
  return address;
}

export const hasAdminOverride = (exportPayload) => {
  const isLandingOverridden = (landing) =>
    landing &&
    landing.model &&
    landing.model.vessel &&
    landing.model.vessel.vesselOverriddenByAdmin;

  return exportPayload &&
    exportPayload.items &&
    exportPayload.items.some(
      (item) =>
        item.landings &&
        item.landings.some((landing) => isLandingOverridden(landing))
    )
    ? true
    : false;
};