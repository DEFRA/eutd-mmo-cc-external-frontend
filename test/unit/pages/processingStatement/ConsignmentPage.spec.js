import * as React from 'react';
import { mount } from 'enzyme/build';
import { Provider } from 'react-redux';
import { Router, Route } from 'react-router-dom';
import { act } from 'react-dom/test-utils';
import { createMemoryHistory } from 'history';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { getProcessingStatementFromRedis, saveProcessingStatementToRedis, clearProcessingStatement} from '../../../../src/client/actions';
import { component as ConsignmentPage } from '../../../../src/client/pages/processingStatement/consignmentPage';
import ConsignmentPageWrapper from '../../../../src/client/pages/processingStatement/consignmentPage';
import * as ErrorUtils from '../../../../src/client/pages/utils/errorUtils';
import * as clientAction from '../../../../src/client/actions/index.js';
import { render } from '@testing-library/react';

jest.mock('../../../../src/client/actions');

describe('Consignment page', () => {
  const mockStore = configureStore([thunk]);
  const mockClearErrors = jest.fn();
  const mockSaveToRedis = jest.spyOn(clientAction, 'saveProcessingStatementToRedis');
  const mockClearProcessingStatement = jest.spyOn(clientAction, 'clearProcessingStatement')
  const mockScrollToErrorIsland = jest.spyOn(ErrorUtils, 'scrollToErrorIsland');
  mockScrollToErrorIsland.mockReturnValue(null);

  beforeEach(() => {
    mockClearErrors.mockReset();
    getProcessingStatementFromRedis.mockReturnValue({
      type: 'GET_PROCESSING_STATEMENT_FROM_REDIS',
    });
    mockClearProcessingStatement.mockReturnValue({ type: 'CLEAR_PROCESSING_STATEMENT '});

    mockScrollToErrorIsland.mockReset();
    mockScrollToErrorIsland.mockImplementation(() => null);
  });

  afterEach(() => {
    saveProcessingStatementToRedis.mockReset();
  });

  it('should redirect to forbidden page when unauthorised', async () => {
    const documentNumber = 'document123';
    const history = createMemoryHistory({
      initialEntries: [
        `/create-processing-statement/${documentNumber}/add-consignment`,
      ],
    });
    const mockPush = jest.spyOn(history, 'push');
    mockPush.mockReturnValue(null);

    const store = mockStore({
      processingStatement: {
        unauthorised: true,
      },
      t: jest.fn(),
    });

    const props = {
      t: jest.fn()
    };

    await act(async () => {
      await mount(
        <Provider store={store}>
          <Router history={history}>
            <Route path="/create-processing-statement/:documentNumber/add-consignment">
              <ConsignmentPage
                route={{
                  previousUri:
                    '/create-processing-statement/:documentNumber/previous-uri',
                    progressUri: '/create-processing-statement/:documentNumber/progress',
                }}

                {...props}
              />
            </Route>
          </Router>
        </Provider>
      );
    });

    expect(mockPush).toHaveBeenCalledWith(
      '/forbidden'
    );
  });

  it('should take a snapshot of the whole page', async () => {
    let  wrapper;
    const documentNumber = 'document123';
    const history = createMemoryHistory({
      initialEntries: [
        `/create-processing-statement/${documentNumber}/add-consignment`,
      ],
    });
    const mockPush = jest.spyOn(history, 'push');
    mockPush.mockReturnValue(null);

    const store = mockStore({
      processingStatement: {
        unauthorised: true,
        errors: null,
        errorsUrl: `/create-processing-statement/${documentNumber}/add-consignment`
      },
      clear: mockClearErrors
    });

    wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <Route path="/create-processing-statement/:documentNumber/add-consignment">
              <ConsignmentPage
                route={{
                  previousUri:
                    '/create-processing-statement/:documentNumber/previous-uri',
                    progressUri: '/create-processing-statement/:documentNumber/progress',
                }}
              />
          </Route>
        </Router>
      </Provider>
    );

    await act(async () => await wrapper);

    const { container } = render(wrapper);
    expect(container).toMatchSnapshot();
  });

  it('when unauthorised is false', async () => {
    let  wrapper;
    const documentNumber = 'document123';
    const history = createMemoryHistory({
      initialEntries: [
        `/create-processing-statement/${documentNumber}/add-consignment`,
      ],
    });
    const mockPush = jest.spyOn(history, 'push');
    mockPush.mockReturnValue(null);

    const store = mockStore({
      processingStatement: {
        unauthorised: false
      }
    });

    wrapper = await mount(
      <Provider store={store}>
        <Router history={history}>
          <Route path="/create-processing-statement/:documentNumber/add-consignment">
              <ConsignmentPage
                route={{
                  previousUri:
                    '/create-processing-statement/:documentNumber/previous-uri',
                    progressUri: '/create-processing-statement/:documentNumber/progress',
                }}
              />
          </Route>
        </Router>
      </Provider>
    );

    await act(async () => await wrapper);
    expect(wrapper).toBeDefined();
  });

  it('should be scroll to errorIsLand', async () => {
    const documentNumber = 'document123';
    const history = createMemoryHistory({
      initialEntries: [
        `/create-processing-statement/${documentNumber}/add-consignment`,
      ],
    });
    const mockPush = jest.spyOn(history, 'push');
    mockPush.mockReturnValue(null);
    mockSaveToRedis.mockReturnValue({
      errors: {
        consignmentDescription: 'Enter a description and commodity code',
      },
    });
    const mockScrollToErrorIsland = jest.spyOn(ErrorUtils, 'scrollToErrorIsland');
    const props = {
      route: {
        path: ':documentNumber/add-consignment-details',
        nextUri: ':documentNumber/add-catch-details',
        previousUri: ':documentNumber/add-exporter-details',
        progressUri: ':documentNumber/progress',
        journey: 'processingStatement',
        title: 'Create a UK processing statement - GOV.UK',
        saveAsDraftUri: 'processing-statements'
      },
      match:{
        params: {documentNumber}
      },
      processingStatement: {
        unauthorised: false,
        errors: {
          consignmentDescription: 'Enter a description and commodity code'
        },
        errorsUrl: `/create-processing-statement/${documentNumber}/add-consignment`
      },
      history: {
        ...history,
        location: {
          pathname: `/create-processing-statement/${documentNumber}/add-consignment`
        }
      },
      clear: mockClearErrors,
      t: jest.fn(),
      saveToRedis: mockSaveToRedis
    };

    await new ConsignmentPage.WrappedComponent({...props}).save('documentNumber/add-catch-weight');

    expect(mockSaveToRedis).toHaveBeenCalledWith({
      data: props.processingStatement || {},
      currentUrl: `/create-processing-statement/${documentNumber}/add-consignment-details`,
      saveToRedisIfErrors: true,
      saveAsDraft: false,
      documentNumber: documentNumber
    });

    expect(mockScrollToErrorIsland).toHaveBeenCalled();
  });

  it('will call all methods needed to load the component', async () => {
    const store = {
      dispatch: () => {
        return new Promise((resolve) => {
          resolve();
        });
      }
    };

    await ConsignmentPageWrapper.loadData(store, 'processingStatement');

    expect(getProcessingStatementFromRedis).toHaveBeenCalled();
  });

  it('will call the the on submit for component', async() => {
    let  wrapper;
    const documentNumber = 'document123';
    const history = createMemoryHistory({
      initialEntries: [
        `/create-processing-statement/${documentNumber}/add-consignment`,
      ],
    });
    const mockPush = jest.spyOn(history, 'push');
    mockPush.mockReturnValue(null);

    const store = mockStore({
      processingStatement: {
        unauthorised: false,
      },
    });

    wrapper = await mount(
      <Provider store={store}>
        <Router history={history}>
          <Route path="/create-processing-statement/:documentNumber/add-consignment">
              <ConsignmentPage
                route={{
                  previousUri:
                    '/create-processing-statement/:documentNumber/previous-uri',
                    progressUri: '/create-processing-statement/:documentNumber/progress',
                }}
              />
          </Route>
        </Router>
      </Provider>
    );

    const event = {
      preventDefault: jest.fn()
    };
    const formEl = wrapper.find('form');
    formEl.props().onSubmit(event);
    expect(formEl.exists()).toBeTruthy();
    expect(event.preventDefault).toHaveBeenCalled();
  });

  it('should clear errors on component will unmount', () => {
    const mockDispatchClearErrors = jest.fn();

    new ConsignmentPage.WrappedComponent({
      match: {
        params: {
          documentNumber: ''
        }
      },
      unauthorised: true,
      clear: mockDispatchClearErrors
    }).componentWillUnmount();

    expect(mockDispatchClearErrors).toHaveBeenCalled();
  });

  it('when processingStatement is undefined', async () => {
    let  wrapper;
    const documentNumber = 'document123';
    const history = createMemoryHistory({
      initialEntries: [
        `/create-processing-statement/${documentNumber}/add-consignment`,
      ],
    });
    const mockPush = jest.spyOn(history, 'push');
    mockPush.mockReturnValue(null);

    const store = mockStore({
      processingStatement: undefined,
    });

    wrapper = await mount(
      <Provider store={store}>
        <Router history={history}>
          <Route path="/create-processing-statement/:documentNumber/add-consignment">
              <ConsignmentPage
                route={{
                  previousUri:
                    '/create-processing-statement/:documentNumber/previous-uri',
                    progressUri: '/create-processing-statement/:documentNumber/progress',
                }}
              />
          </Route>
        </Router>
      </Provider>
    );

    await act(async () => await wrapper);
    expect(wrapper).toBeDefined();
  });

  it('should call on onChange/goBack when entering a description', async () => {
    const documentNumber = 'document123';
    const history = createMemoryHistory({
      initialEntries: [
        `/create-processing-statement/${documentNumber}/add-consignment`,
      ],
    });
    const mockPush = jest.spyOn(history, 'push');
    mockPush.mockReturnValue(null);
    mockSaveToRedis.mockReturnValue({
      errors: {
        consignmentDescription: 'Enter a description and commodity code',
      },
    });

    const props = {
      route: {
        path: ':documentNumber/add-consignment-details',
        nextUri: ':documentNumber/add-catch-details',
        previousUri: ':documentNumber/add-exporter-details',
        progressUri: ':documentNumber/progress',
        journey: 'processingStatement',
        title: 'Create a UK processing statement - GOV.UK',
        saveAsDraftUri: 'processing-statements'
      },
      match:{
        params: {documentNumber}
      },
      processingStatement: undefined,
      history: {
        ...history,
        location: {
          pathname: `/create-processing-statement/${documentNumber}/add-consignment`
        }
      },
      clear: mockClearErrors,
      t: jest.fn(),
      saveToRedis: mockSaveToRedis,
      save: jest.fn()
    };

    await new ConsignmentPage.WrappedComponent({...props}).onChange({
      target:{
        name: 'consignmentDescription'
      }
    });

    await new ConsignmentPage.WrappedComponent({...props}).goBack({
      preventDefault: jest.fn()
    });

    expect(props.save).toHaveBeenCalled();
  });
});

describe('#Consignment page with translation', () => {
  let wrapper;
  const documentNumber = 'document123';
  const mockStore = configureStore([thunk]);
  const mockClearErrors = jest.fn();
  let mockScrollToErrorIsland;
  let mockSaveToRedis;

  const processingStatement = {
    unauthorised: true,
    errors: {
      consignmentDescription: 'psConsignmentEnterConsignmentDescription'
    },
    errorsUrl: `/create-processing-statement/${documentNumber}/add-consignment`
  };

  const props = {
    route: {
      previousUri:
        '/create-processing-statement/:documentNumber/previous-uri',
      progressUri: '/create-processing-statement/:documentNumber/progress',
      saveAsDraftUri: ':documentNumber/add-catch-details',
      nextUri: ':documentNumber/add-catch-details'
    },
    t: jest.fn(),
    saveToRedis: mockSaveToRedis
  };

  beforeEach(() => {
    mockClearErrors.mockReset();
    getProcessingStatementFromRedis.mockReturnValue({
      type: 'GET_PROCESSING_STATEMENT_FROM_REDIS',
    });

    mockScrollToErrorIsland = jest.spyOn(ErrorUtils, 'scrollToErrorIsland');
    mockSaveToRedis = jest.spyOn(clientAction, 'saveProcessingStatementToRedis');

    clearProcessingStatement.mockReturnValue({ type:  'CLEAR_PROCESSING_STATEMENT' })
  });

  afterEach(() => {
    saveProcessingStatementToRedis.mockReset();
    mockScrollToErrorIsland.mockRestore();
    mockSaveToRedis.mockRestore();
  });

  const getConsignmentWrapper = async() => {
    const history = createMemoryHistory({
      initialEntries: [
        `/create-processing-statement/${documentNumber}/add-consignment`,
      ],
    });
    const mockPush = jest.spyOn(history, 'push');
    mockPush.mockReturnValue(null);

    const store = mockStore({
      processingStatement,
      t: jest.fn(),
    });

    await act(async () => {
      wrapper = await mount(
        <Provider store={store}>
          <Router history={history}>
            <Route path="/create-processing-statement/:documentNumber/add-consignment">
              <ConsignmentPage
                {...props}
              />
            </Route>
          </Router>
        </Provider>
      );
    });
    return wrapper
  }

  it('should translate to welsh', async () => {
    await getConsignmentWrapper();
    wrapper.find('button#continue').simulate('click');
    mockSaveToRedis.mockReturnValue({
      errors: {
        consignmentDescription: 'psConsignmentEnterConsignmentDescription',
      },
    });

    expect(mockSaveToRedis).toHaveBeenCalledWith({
      data: processingStatement || {},
      currentUrl: `/create-processing-statement/${documentNumber}/add-consignment-details`,
      saveToRedisIfErrors: true,
      saveAsDraft: true,
      documentNumber: documentNumber
    });

    expect(wrapper.find('ErrorIsland h2.error-summary-heading').text()).toEqual('There is a problem');
    expect(wrapper.find('ErrorIsland ul.error-summary-list li').at(0).text()).toEqual('Enter a description and commodity code');
    expect(wrapper.find('TextArea span').at(2).text()).toBe('Enter a description and commodity code');
  });
});
