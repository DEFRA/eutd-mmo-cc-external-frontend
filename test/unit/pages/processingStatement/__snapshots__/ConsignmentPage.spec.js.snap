// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Consignment page should take a snapshot of the whole page 1`] = `
<div>
  <div
    class="processing-statement css-1l99ixu-OuterContainer e4u1u2g0"
  >
    <div
      class="css-n57fe4-InnerContainer e4u1u2g1"
    >
      <div
        class="css-1lzr9ic-StyledRow e4u1u2g0"
      >
        <div
          class="css-1w80l5h-StyledColumn e4u1u2g0"
        >
          <a
            class="e1vcssft0 css-zk98nb-Anchor-StyledHoc e4u1u2g0"
            href="/create-processing-statement/document123/previous-uri"
          >
            Back
          </a>
          <h1
            class="e1vcssft0 css-11nzzut-StyledHeader-StyledHoc e4u1u2g0"
          >
            Add a description and commodity code
          </h1>
        </div>
      </div>
      <form
        action="/orchestration/api/v1/processingStatement/saveAndValidate"
        method="POST"
      >
        <div
          class="css-1lzr9ic-StyledRow e4u1u2g0"
        >
          <div
            class="css-n58wva-StyledColumn e4u1u2g0"
          >
            <label
              class="e1vcssft0 css-137pbp-StyledLabel-StyledHoc e4u1u2g0"
              for="consignmentDescription"
            >
              <span
                class="e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
              />
              <span
                class="e1vcssft0 css-1whl0n4-StyledHint-StyledHoc e4u1u2g0"
              >
                You need to tell us what’s in the consignment and include the relevant commodity codes. For example, ‘Herring fillets (16041210), Atlantic cod fishcakes (16041992)’.
              </span>
              
              <textarea
                autocomplete="off"
                class="formControl css-1dvz68f-TextAreaField e4u1u2g0"
                id="consignmentDescription"
                name="consignmentDescription"
                rows="5"
                type="text"
              />
            </label>
          </div>
        </div>
        <div
          class="css-1lzr9ic-StyledRow e4u1u2g0"
        >
          <button
            class="css-17uw1q2-StyledButton-StyledHoc e4u1u2g0"
            formaction="/orchestration/api/v1/processingStatement/saveAndValidate?c=undefined&n=undefined&saveAsDraftUrl=undefined"
            id="saveAsDraft"
            name="saveAsDraft"
            type="submit"
          >
            Save as Draft
          </button>
          <input
            name="journey"
            type="hidden"
            value=""
          />
          <input
            name="currentUri"
            type="hidden"
            value=""
          />
          <button
            class="css-1gc9650-StyledButton-StyledHoc e4u1u2g0"
            id="continue"
          >
             
            Save and continue
          </button>
        </div>
      </form>
      <a
        href="/create-processing-statement/document123/progress"
        id="backToProgress"
      >
        Back to Your Progress
        <span
          class="govuk-visually-hidden"
        >
          back to your progress page link
        </span>
      </a>
      <div
        class="css-1lzr9ic-StyledRow e4u1u2g0"
      >
        <div
          class="css-1w80l5h-StyledColumn e4u1u2g0"
        >
          <hr
            class="help-link"
          />
          <h2
            class="e1vcssft0 css-162lx7h-StyledHeader-StyledHoc e4u1u2g0"
          >
            Need help?
          </h2>
          <a
            href="undefined"
            rel="noopener noreferrer"
            target="_blank"
          >
            Get help exporting fish from the UK (gov.uk)
             
            <span
              class="govuk-visually-hidden"
            >
              (opens in new tab)
            </span>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
